openem-ingestor/desktop-app/app.go

```diff
@@ -0,0 +1,67 @@
+package main
+
+import (
+	"context"
+	"openem-ingestor/core"
+
+	"github.com/google/uuid"
+	"github.com/wailsapp/wails/v2/pkg/runtime"
+)
+
+// App struct
+type App struct {
+	ctx       context.Context
+	taskqueue core.TaskQueue
+	config    core.Config
+}
+
+// NewApp creates a new App application struct
+func NewApp(config core.Config) *App {
+	return &App{config: config, taskqueue: core.TaskQueue{Config: config}}
+}
+
+// Show prompt before closing the app
+func (b *App) beforeClose(ctx context.Context) (prevent bool) {
+	dialog, err := runtime.MessageDialog(ctx, runtime.MessageDialogOptions{
+		Type:    runtime.QuestionDialog,
+		Title:   "Quit?",
+		Message: "Are you sure you want to quit? This will stop all pending downloads.",
+	})
+
+	if err != nil {
+		return false
+	}
+	return dialog != "Yes"
+}
+
+// startup is called when the app starts. The context is saved
+// so we can call the runtime methods
+func (a *App) startup(ctx context.Context) {
+	a.ctx = ctx
+	a.taskqueue.AppContext = a.ctx
+	a.taskqueue.Startup()
+}
+
+func (a *App) SelectFolder() {
+	folder, err := core.SelectFolder(a.ctx)
+	if err != nil {
+		return
+	}
+
+	err = a.taskqueue.CreateTask(folder)
+	if err != nil {
+		return
+	}
+}
+
+func (a *App) CancelTask(id string) {
+	a.taskqueue.CancelTask(uuid.MustParse(id))
+}
+func (a *App) RemoveTask(id string) {
+	a.taskqueue.RemoveTask(uuid.MustParse(id))
+}
+
+func (a *App) ScheduleTask(id uuid.UUID) {
+
+	a.taskqueue.ScheduleTask(id)
+}
```

Comment:
very self explanatory, once you understand the task system a little! I like this

Reply:
I think at some point we need to think about how persistent these tasks should be; currently everything is stored in memory. The tasks will not survive any restarts. There are proper task queue implementations that allow storing task/results in databases (can also just be a file) but also for the service use case it might need to be a bit more reliable.

Reply:
It would be better to save them to some persistent storage, but it's not a priority for now. If the transfer never completes due to the service crashing or restarting, they can re-request ingestion, maybe with an option to use a pre-existing entry in the scicat database (or use the sourceFolder to see if there's already one).